Sources/PrivacyStats/PrivacyStats.swift

```diff
@@ -0,0 +1,268 @@
+                }
+
+                // Check if the pack we're currently storing is from a previous day.
+                let isCurrentDayPack = pack.timestamp == Date.currentPrivacyStatsPackTimestamp
+
+                do {
+                    let statsObjects = try PrivacyStatsUtils.fetchOrInsertCurrentStats(for: Set(pack.trackers.keys), in: context)
+                    statsObjects.forEach { stats in
+                        if let count = pack.trackers[stats.companyName] {
+                            stats.count = count
+                        }
+                    }
+
+                    // When storing a pack from a previous day, we may have outdated packs, so delete them as needed.
+                    if !isCurrentDayPack {
+                        PrivacyStatsUtils.deleteOutdatedPacks(in: context)
+                    }
+
+                    guard context.hasChanges else {
+                        continuation.resume()
+                        return
+                    }
+
+                    try context.save()
+                    Logger.privacyStats.debug("Saved stats \(pack.timestamp) \(pack.trackers)")
+
+                    // Only emit update event when saving current-day pack. For previous-day pack,
+                    // a follow-up commit event will come and we'll emit the update then.
+                    if isCurrentDayPack {
+                        statsUpdateSubject.send()
+                    }
+                } catch {
+                    Logger.privacyStats.error("Save error: \(error)")
+                    errorEvents?.fire(.failedToStorePrivacyStats(error))
+                }
+                continuation.resume()
+            }
+        }
+    }
+
+    /**
+     * This function is only called in the initializer. It performs a blocking call to the database
+     * to spare us the hassle of declaring the initializer async or spawning tasks from within the
+     * initializer without being able to await them, thus making testing trickier.
+     */
+    private func initializeCurrentPack() -> PrivacyStatsPack {
+        var pack: PrivacyStatsPack?
+        context.performAndWait {
+            let timestamp = Date.currentPrivacyStatsPackTimestamp
+            do {
+                let currentDayStats = try PrivacyStatsUtils.loadCurrentDayStats(in: context)
+                Logger.privacyStats.debug("Loaded stats \(timestamp) \(currentDayStats)")
+                pack = PrivacyStatsPack(timestamp: timestamp, trackers: currentDayStats)
+
+                PrivacyStatsUtils.deleteOutdatedPacks(in: context)
+                try context.save()
+            } catch {
+                Logger.privacyStats.error("Faild to load current stats: \(error)")
+                errorEvents?.fire(.failedToLoadCurrentPrivacyStats(error))
+            }
+        }
+        return pack ?? PrivacyStatsPack(timestamp: Date.currentPrivacyStatsPackTimestamp)
+    }
+
+    private func subscribeToAppTermination() {
```

Comment:
I understand why it’s a nice approach to encapsulate this login here, as the client does not need to worry about reacting to `willTerminate` events. However, two thoughts:

1. I’m unsure if the `PrivacyStats` “layer" should be aware of app events (and importing `AppKit` and `UIKit`). 
2. The same way other `PrivacyStats` actions are explicit public methods (e.g recording a tracker), maybe committing changes should be an explicit action, which the client needs to be conscious of. At the moment it feels like a side effect which is happening, hidden from the clients

So, one suggestion would be to make this a public function which clients need to call in their own `willTerminate` methods.

Reply:
I appreciate you questioning this and I agree it feels off. The reason I included it here was that HistoryTabExtension does a similar thing but that one is fully in the macOS browser. Let me move it away from BSK and only leave the function here to allow saving the state on demand. I'll also remove the RunLoop thingy (also copied over from another place in the app) because this isn't some crucial data that we're storing here :)

Reply:
I need to keep the RunLoop around anyway because it's an asynchronous function. But just so you know, it's a common pattern used in the app and looks safe to me.